rs/xnet/payload_builder/src/lib.rs

```diff
@@ -910,6 +926,24 @@ pub fn get_msg_limit(subnet_id: SubnetId, state: &ReplicatedState) -> Option<usi
     }
 }
 
+/// Calculates an upper bound for how many messages can be included into a block based on how many
+/// signals the reverse stream will contain after inducting the stream slice.
+///
+/// This is computed such that the stream index of the last message is at most `slice_begin +
+/// 50_000`. Since `slice_begin` is used for garbage collecting signals, this ensures at most
+/// `50_000` signals in the reverse stream.
+pub fn get_signal_limit(begin: Option<&ExpectedIndices>, slice_begin: StreamIndex) -> usize {
+    // TODO: Replace the 50_000 with a constant. This should be the same constant used in the
+    // stream builder: `MAX_STREAM_MESSAGES`.
+    begin.map_or(50_000, |begin| {
+        let slice_end_index = slice_begin + 50_000.into();
```

Comment:
Nit: I'd call this
```suggestion
        let max_slice_end_index = slice_begin + 50_000.into();
```
so that it is clear that this is the last index we would possibly want to accept.